stdlib/src/builtin/string.mojo

```diff
@@ -388,6 +388,134 @@ fn atol(str: String, base: Int = 10) raises -> Int:
+    as an float value, an error is raised. For example, `atof("hi")` raises an
+    error.
+
+    Args:
+        str_ref: A string to be parsed as a floating point.
+
+    Returns:
+        An float value that represents the string, or otherwise raises.
+    """
+    if not str_ref:
+        raise Error(_atof_error(str_ref))
+
+    var result: Float64 = 0.0
+    var exponent: Int = 0
+    var is_negative: Bool = False
+
+    alias ord_0 = ord("0")
+    alias ord_9 = ord("9")
+    alias ord_dot = ord(".")
+    alias ord_plus = ord("+")
+    alias ord_minus = ord("-")
+    alias ord_f = ord("f")
+    alias ord_F = ord("F")
+    alias ord_e = ord("e")
+    alias ord_E = ord("E")
+
+    var start: Int = 0
+    var str_len = len(str_ref)
+    var buff = str_ref.unsafe_ptr()
+
+    var c: Int = 0
+    # skip leading spaces and read sign
+    for pos in range(start, str_len):
+        c = int(buff[pos])
+        if isspace(c):
+            pass
+        elif c == ord_minus:
+            if is_negative:
+                raise Error(_atof_error(str_ref))
+            is_negative = True
+        else:
+            break
+        start += 1
+    # read before dot
+    for pos in range(start, str_len):
+        c = int(buff[pos])
+        if ord_0 <= c <= ord_9:
+            result = result * 10.0 + (c - ord_0)
+            start += 1
+        else:
+            break
+    # if dot -> read after dot
+    if c == ord_dot:
+        start += 1
+        for pos in range(start, str_len):
+            c = int(buff[pos])
+            if ord_0 <= c <= ord_9:
+                result = result * 10.0 + (c - ord_0)
+                exponent -= 1
+            else:
+                break
+            start += 1
+    # if e/E -> read scientific notation
+    if c == ord_e or c == ord_E:
+        start += 1
+        var sign: Int = 1
+        var shift: Int = 0
+        var has_number: Bool = False
+        for pos in range(start, str_len):
+            c = int(buff[pos])
+            if c == ord_plus:
+                pass
+            elif c == ord_minus:
+                sign = -1
+            elif ord_0 <= c <= ord_9:
+                has_number = True
+                shift = shift * 10 + (c - ord_0)
+            else:
+                break
+            start += 1
+        exponent += sign * shift
+        if not has_number:
+            raise Error(_atof_error(str_ref))
+    # check if the string is fully parsed
+    if start != str_len:
+        if c == ord_f or c == ord_F:  # f/F at end is allowed
+            start += 1
+        for pos in range(start, str_len):
+            c = int(buff[pos])
+            if not isspace(c):
+                raise Error(_atof_error(str_ref))
+    # apply shift
+    var shift: Int = 10 ** abs(exponent)
+    if exponent > 0:
+        result *= shift
+    if exponent < 0:
+        result /= shift
+    # apply sign
+    if is_negative:
+        result = -result
+    return result
```

Comment:
style nit:
```suggestion
    return -result if is_negative else result
```
Alternatively, a branchless implementation might be even better: replace `is_negative` with `var sign: Int = 1` and just `return sign * result` in the end.